File: desktop/src/shared/theme/adaptive-theme.ts
/**
 * Adaptive Theme Engine
 *
 * Derives shadcn CSS variables from a syntax theme's key colors (bg, fg, comment, git).
 * Detects light vs dark from background luminance and adjusts accordingly.
 *
 * Ported from builderbot/apps/staged/src/lib/theme.ts, flattened to emit
 * shadcn CSS vars directly (no intermediate Theme object).
 */

// =============================================================================
// Color Utilities
// =============================================================================

interface RGB {
  r: number;
  g: number;
  b: number;
}

function hexToRgb(hex: string): RGB {
  const long = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})?$/i.exec(
    hex,
  );
  if (long) {
    return {
      r: parseInt(long[1], 16),
      g: parseInt(long[2], 16),
      b: parseInt(long[3], 16),
    };
  }

  const short = /^#?([a-f\d])([a-f\d])([a-f\d])([a-f\d])?$/i.exec(hex);
  if (short) {
    return {
      r: parseInt(short[1] + short[1], 16),
      g: parseInt(short[2] + short[2], 16),
      b: parseInt(short[3] + short[3], 16),
    };
  }

  return { r: 128, g: 128, b: 128 };
}

function rgbToHex({ r, g, b }: RGB): string {
  const clamp = (n: number) => Math.max(0, Math.min(255, Math.round(n)));
  return `#${[r, g, b].map((c) => clamp(c).toString(16).padStart(2, "0")).join("")}`;
}

export function luminance(hex: string): number {
  const { r, g, b } = hexToRgb(hex);
  const [rs, gs, bs] = [r, g, b].map((c) => {
    const s = c / 255;
    return s <= 0.03928 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4;
  });
  return 0.2126 * rs + 0.7152 * gs + 0.0722 * bs;
}

function mix(hex1: string, hex2: string, factor: number): string {
  const c1 = hexToRgb(hex1);
  const c2 = hexToRgb(hex2);
  return rgbToHex({
    r: c1.r + (c2.r - c1.r) * factor,
    g: c1.g + (c2.g - c1.g) * factor,
    b: c1.b + (c2.b - c1.b) * factor,
  });
}

function adjust(hex: string, amount: number): string {
  const target = amount > 0 ? "#ffffff" : "#000000";
  return mix(hex, target, Math.abs(amount));
}

function overlay(hex: string, alpha: number): string {
  const { r, g, b } = hexToRgb(hex);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}

// =============================================================================
// Chrome Color Calculation
// =============================================================================

const CONTRAST_VALUE = 0.035;
const CONTRAST_OFFSET = 0.0135;

function calculateLumDiff(bgLum: number): number {
  return CONTRAST_VALUE * Math.log(1 + (bgLum + CONTRAST_OFFSET) * 10);
}

function findColorWithLuminance(baseColor: string, targetLum: number): string {
  const baseLum = luminance(baseColor);
  if (Math.abs(baseLum - targetLum) < 0.001) return baseColor;

  const target = targetLum < baseLum ? "#000000" : "#ffffff";
  let lo = 0;
  let hi = 1;

  for (let i = 0; i < 20; i++) {
    const mid = (lo + hi) / 2;
    const testLum = luminance(mix(baseColor, target, mid));
    const diff = testLum - targetLum;

    if (Math.abs(diff) < 0.001) break;

    if (target === "#000000") {
      if (testLum > targetLum) lo = mid;
      else hi = mid;
    } else {
      if (testLum < targetLum) lo = mid;
      else hi = mid;
    }
  }
  return mix(baseColor, target, (lo + hi) / 2);
}

function calculateChromeColors(syntaxBg: string): {
  chrome: string;
  primary: string;
} {
  const bgLum = luminance(syntaxBg);
  const lumDiff = calculateLumDiff(bgLum);
  const targetChromeLum = bgLum - lumDiff;

  if (targetChromeLum >= 0) {
    return {
      chrome: findColorWithLuminance(syntaxBg, targetChromeLum),
      primary: syntaxBg,
    };
  }

  return {
    chrome: findColorWithLuminance(syntaxBg, 0),
    primary: findColorWithLuminance(syntaxBg, lumDiff),
  };
}

// =============================================================================
// Hex → HSL component format ("H S% L%") for Tailwind's hexToHsl() wrappers
// =============================================================================

export function hexToHsl(hex: string): string {
  const { r, g, b } = hexToRgb(hex);
  const rn = r / 255;
  const gn = g / 255;
  const bn = b / 255;

  const max = Math.max(rn, gn, bn);
  const min = Math.min(rn, gn, bn);
  const l = (max + min) / 2;

  if (max === min) {
    return `0 0% ${(l * 100).toFixed(1)}%`;
  }

  const d = max - min;
  const s = l > 0.5 ? d / (2 - max - min) : d / (max + min);

  let h: number;
  if (max === rn) {
    h = ((gn - bn) / d + (gn < bn ? 6 : 0)) / 6;
  } else if (max === gn) {
    h = ((bn - rn) / d + 2) / 6;
  } else {
    h = ((rn - gn) / d + 4) / 6;
  }

  return `${(h * 360).toFixed(1)} ${(s * 100).toFixed(2)}% ${(l * 100).toFixed(1)}%`;
}

// =============================================================================
// Adaptive Theme Generator — emits shadcn CSS vars directly
// =============================================================================

export interface ThemeGitColors {
  added: string | null;
  deleted: string | null;
  modified: string | null;
}

export interface ThemeResult {
  isDark: boolean;
  vars: Record<string, string>;
}

/**
 * Derive a full set of shadcn CSS variables from syntax theme colors.
 *
 * Takes bg, fg, comment hex colors (+ optional git decoration colors) and
 * returns the var map ready to apply via style.setProperty().
 */
export function createThemeVars(
  syntaxBg: string,
  syntaxFg: string,
  syntaxComment: string,
  gitColors?: ThemeGitColors,
): ThemeResult {
  const isDark = luminance(syntaxBg) < 0.5;

  const { chrome: chromeColor, primary: primaryBg } =
    calculateChromeColors(syntaxBg);

  const dir = isDark ? 1 : -1;
  const elevate = (amount: number) => adjust(primaryBg, dir * amount);

  // Git/accent colors with fallbacks
  const fallbackGreen = isDark ? "#3fb950" : "#1a7f37";
  const fallbackRed = isDark ? "#f85149" : "#cf222e";
  const fallbackOrange = isDark ? "#d29922" : "#9a6700";

  const accentGreen = gitColors?.added ?? fallbackGreen;
  const accentRed = gitColors?.deleted ?? fallbackRed;
  const accentOrange = fallbackOrange;

  // Derived colors
  const borderColor = mix(primaryBg, syntaxFg, isDark ? 0.15 : 0.12);
  const hoverBg = elevate(0.06);
  const primaryFg = hexToHsl(primaryBg);
  const textFg = hexToHsl(syntaxFg);

  return {
    isDark,
    vars: {
      // Backgrounds
      "--background": hexToHsl(primaryBg),
      "--card": hexToHsl(primaryBg),
      "--popover": hexToHsl(elevate(0.08)),
      "--muted": hexToHsl(hoverBg),
      "--accent": hexToHsl(hoverBg),
      "--secondary": hexToHsl(hoverBg),

      // Foregrounds
      "--foreground": textFg,
      "--card-foreground": textFg,
      "--popover-foreground": textFg,
      "--muted-foreground": hexToHsl(syntaxComment),
      "--accent-foreground": textFg,
      "--secondary-foreground": textFg,

      // Destructive
      "--destructive": hexToHsl(accentRed),
      "--destructive-foreground": primaryFg,

      // Borders
      "--border": hexToHsl(borderColor),
      "--input": hexToHsl(borderColor),
      "--ring": textFg,

      // Sidebar
      "--sidebar-background": hexToHsl(chromeColor),
      "--sidebar-foreground": textFg,
      "--sidebar-accent": hexToHsl(primaryBg),
      "--sidebar-accent-foreground": textFg,
      "--sidebar-border": hexToHsl(borderColor),
      "--sidebar-ring": hexToHsl(borderColor),

      // Status colors (hex — used directly via var())
      "--status-added": accentGreen,
      "--status-deleted": accentRed,
      "--status-modified": accentOrange,

      // Warning
      "--ui-warning": accentOrange,
      "--ui-warning-bg": overlay(accentOrange, isDark ? 0.1 : 0.08),
    },
  };
}

File: mobile/lib/shared/theme/adaptive_theme.dart
// Adaptive Theme Engine
//
// Derives a Material 3 [ColorScheme] from a syntax theme's key colors
// (bg, fg, comment, git). Detects light vs dark from background luminance
// and adjusts accordingly.
//
// Ported from desktop/src/shared/theme/adaptive-theme.ts.
import 'dart:math' as math;
import 'package:flutter/material.dart';

import 'color_scheme.dart' show contrastForeground;
import 'theme_catalog.dart';

// ---------------------------------------------------------------------------
// Color utilities
// ---------------------------------------------------------------------------

double _luminance(Color c) => c.computeLuminance();

int _to255(double v) => (v * 255.0).round().clamp(0, 255);

Color _mix(Color c1, Color c2, double factor) {
  final r = c1.r + (c2.r - c1.r) * factor;
  final g = c1.g + (c2.g - c1.g) * factor;
  final b = c1.b + (c2.b - c1.b) * factor;
  return Color.fromARGB(255, _to255(r), _to255(g), _to255(b));
}

Color _adjust(Color c, double amount) {
  final target = amount > 0 ? const Color(0xFFFFFFFF) : const Color(0xFF000000);
  return _mix(c, target, amount.abs());
}

// ---------------------------------------------------------------------------
// Chrome color calculation (sidebar/chrome area, slightly darker than editor)
// ---------------------------------------------------------------------------

const _contrastValue = 0.035;
const _contrastOffset = 0.0135;

double _calculateLumDiff(double bgLum) {
  return _contrastValue * math.log(1 + (bgLum + _contrastOffset) * 10);
}

Color _findColorWithLuminance(Color base, double targetLum) {
  final baseLum = _luminance(base);
  if ((baseLum - targetLum).abs() < 0.001) return base;

  final target = targetLum < baseLum
      ? const Color(0xFF000000)
      : const Color(0xFFFFFFFF);
  var lo = 0.0;
  var hi = 1.0;

  for (var i = 0; i < 20; i++) {
    final mid = (lo + hi) / 2;
    final testLum = _luminance(_mix(base, target, mid));
    final diff = testLum - targetLum;

    if (diff.abs() < 0.001) break;

    if (target == const Color(0xFF000000)) {
      if (testLum > targetLum) {
        lo = mid;
      } else {
        hi = mid;
      }
    } else {
      if (testLum < targetLum) {
        lo = mid;
      } else {
        hi = mid;
      }
    }
  }
  return _mix(base, target, (lo + hi) / 2);
}

({Color chrome, Color primary}) _calculateChromeColors(Color syntaxBg) {
  final bgLum = _luminance(syntaxBg);
  final lumDiff = _calculateLumDiff(bgLum);
  final targetChromeLum = bgLum - lumDiff;

  if (targetChromeLum >= 0) {
    return (
      chrome: _findColorWithLuminance(syntaxBg, targetChromeLum),
      primary: syntaxBg,
    );
  }

  return (
    chrome: _findColorWithLuminance(syntaxBg, 0),
    primary: _findColorWithLuminance(syntaxBg, lumDiff),
  );
}

// ---------------------------------------------------------------------------
// Public API: generate a ColorScheme from theme colors
// ---------------------------------------------------------------------------

/// Generate a full Material 3 [ColorScheme] from syntax theme colors.
///
/// Maps the desktop's CSS variable system to Material 3 semantic slots:
///   --background     → surface
///   --foreground     → onSurface
///   --muted          → surfaceContainerHighest
///   --muted-fg       → onSurfaceVariant
///   --border         → outline
///   --popover        → surfaceContainerHigh (elevated surfaces)
///   --destructive    → error
///   --sidebar-bg     → surfaceContainerLowest (chrome)
ColorScheme generateColorScheme(ThemeColors theme) {
  final isDark = theme.isDark;
  final syntaxBg = theme.bg;
  final syntaxFg = theme.fg;
  final syntaxComment = theme.comment;

  final (:chrome, :primary) = _calculateChromeColors(syntaxBg);

  final dir = isDark ? 1.0 : -1.0;
  Color elevate(double amount) => _adjust(primary, dir * amount);

  // Fallback git/accent colors
  final fallbackGreen = isDark
      ? const Color(0xFF3FB950)
      : const Color(0xFF1A7F37);
  final fallbackRed = isDark
      ? const Color(0xFFF85149)
      : const Color(0xFFCF222E);

  final accentGreen = theme.added ?? fallbackGreen;
  final accentRed = theme.deleted ?? fallbackRed;

  // Derived surfaces
  final borderColor = _mix(primary, syntaxFg, isDark ? 0.15 : 0.12);
  final hoverBg = elevate(0.06);
  final popoverBg = elevate(0.08);

  return ColorScheme(
    brightness: isDark ? Brightness.dark : Brightness.light,

    // Primary — use the theme fg as a muted "primary" to keep the theme
    // feeling cohesive. Accent color override happens in applyAccent().
    primary: syntaxFg,
    onPrimary: contrastForeground(syntaxFg),
    primaryContainer: hoverBg,
    onPrimaryContainer: syntaxFg,

    // Secondary
    secondary: syntaxComment,
    onSecondary: contrastForeground(syntaxComment),
    secondaryContainer: hoverBg,
    onSecondaryContainer: syntaxFg,

    // Tertiary
    tertiary: accentGreen,
    onTertiary: contrastForeground(accentGreen),
    tertiaryContainer: hoverBg,
    onTertiaryContainer: accentGreen,

    // Error / destructive
    error: accentRed,
    onError: contrastForeground(accentRed),
    errorContainer: _mix(primary, accentRed, 0.15),
    onErrorContainer: accentRed,

    // Surfaces
    surface: primary,
    onSurface: syntaxFg,
    onSurfaceVariant: syntaxComment,

    // Outline / borders
    outline: borderColor,
    outlineVariant: _mix(primary, borderColor, 0.5),

    // Inverse
    inverseSurface: syntaxFg,
    onInverseSurface: primary,
    inversePrimary: syntaxComment,

    // Shadow / scrim
    shadow: const Color(0xFF000000),
    scrim: const Color(0xFF000000),
    surfaceTint: syntaxFg,

    // Container hierarchy
    surfaceContainerLowest: chrome,
    surfaceContainerLow: _mix(chrome, primary, 0.5),
    surfaceContainer: primary,
    surfaceContainerHigh: popoverBg,
    surfaceContainerHighest: elevate(0.04),
  );
}
